### DISCUSSION TITLE
Received SIGTERM. Terminating subprocesses

### DISCUSSION THREAD

**Discussion by SolinTao** [apache/airflow#23806] [General]
### Apache Airflow version

2.2.4

### What happened

My program flow is as follows:
Step1： parent process ——> Step2：start three child processes ( use multiprocessing.Process ) ——> Step3：wait for the child process to finish.
The following error message appears in step 3：
![image](https://user-images.githubusercontent.com/26808654/168945788-29e9939c-4058-4631-acce-8c5746496dab.png)

Why does the process receive an exception signal when subprocess terminates, and how can I resolve it? 

### Operating System

Debian GNU/Linux 11 (bullseye)

### Versions of Apache Airflow Providers

apache-airflow-providers-celery==2.1.0
apache-airflow-providers-ftp=2.0.1
apache-airflow-providers-http=2.0.3
apache-airflow-providers-imap==2.2.0
apache-airflow-providers-mysql==2.2.0 (I use pymysql)
apache-airflow-providers-redis==2.0.1
apache-airflow-providers-sqlite==2.1.0

### Deployment

Other Docker-based deployment

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by pingzh** [comment]
hi @SolinTao what's this `parent process`? is this your airflow task?

also it will be great if you can include more log message before my selected area

![image](https://user-images.githubusercontent.com/8662365/168951941-d0317f4d-a18f-48f7-852e-a6a836c3b66b.png)

**Comment by SolinTao** [comment]
hi @pingzh ，parent process is my airflow task. There is no additional useful log information except the task log. I'm sorry I didn't make it clear,  I used [pandarallel ](https://pypi.org/project/pandarallel/)package in my program. I delete ‘multiprocessing.Process‘ and just keep pandarallel, but there was still an error message. 
`{taskinstance.py:1413} ERROR - Received SIGTERM. Terminating subprocesses`

By the way, is Airflow allowed to use multiple cpus by default in multiple processes? (I didn't change the default configuration)
`[operators] 

default_owner = airflow 
default_cpus = 1
default_ram = 512
default_disk = 512
default_gpus = 0`

**Reply by pingzh** [reply]
it looks like you are using `CeleryExecutor`. if you did not modify the airflow.cfg about the `task_runner`, you can use multiprocessing.Process, as the default StandardTaskRunner does not enforce resources. But if you deploy your celery to k8s, k8s is likely to enforce some resource limit on your task pod.

from the log, it looks like the airflow task process receives `SIGTERM` from external and then the signal_handler kicked in to properly handle the signal.

**Reply by pingzh** [reply]
@potiuk i also expect to see a log message from LocalTaskJob, https://github.com/apache/airflow/blob/2.2.4/airflow/jobs/local_task_job.py#L82 . however, from his log, looks like the message was from https://github.com/apache/airflow/blob/2.2.4/airflow/models/taskinstance.py#L1413 .

@SolinTao can you attach the full log if possible? thanks

**Comment by potiuk** [comment]
It does not look like airflow issue "as such" - there is nothing preventing anyone from spawning multiple processes in airflow - airflow and workers are pretty typical python process - and any "prevention" you might see can only come from the way how you deploy it - you could use cgroups or whatnot - but certainly it's not airlfow limitation.

The sigterm that are coming are coming from somewhere. simply something is sending SIGTERM to your processes. But what that something might-be - hard to say without knowing your deployment (and likely only you can say it by a deeper investigation in your deployment - checking celery or kubernetes logs for details. If there is nothing in Airlfow logs, then likely it's not airflow that does it - but something else. 

Unfortunately you have not provided any details of your deployment, so it is hard for me to try to help you to guide your further investigations, but there might be few wild-guesses:

* if you are using celery, celery internaly uses billiard which is a fork of multiprocessing - and they recommend anyone who tries to start multiprocessing from within Celery tasks is recommended to use it to (import billiard as multiprocessing) - Celery made some modifications that **might** make multi-processing programs to not work correctly if they use "standard" multiprocessing. You can search for a number of issues there just look for "celery billiard" 

* also if you are using celery, you might have different ways of running celery that might actually clash with the ways you run your multi-prorcessing code, specifically you should look at pool option https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#pool - you need to take a close look at the documentation of celery and see if it might clash with your usage.

* another possibility (applicable with many deployments) and very likely one that when spawning your multiple processes you simply lack some resources and your deployment (for example Kubernetes) might start killing pods/containers. same with other deployments. Airflow does not manage resources on its own- so if external deployment lacks resources it's the responsibility of the deployment (and it might start kiling stuff). 

Those are the pointers I might give you, but I am afraid if the logs of Airlfow do not tell the cause, you need to find it by looking in the logs of your deployment,

BTW. I converted it into a discussion. This is not a reproducible airflow issue (unless your investigation will lead us to that) so this is at most a discussion, not a bug.
